wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/OptionReader.kt

```diff
@@ -0,0 +1,183 @@
+    } else if (c != keyValueSeparator) {
+      throw reader.unexpected("expected '$keyValueSeparator' in option")
+    }
+    val kindAndValue = readKindAndValue()
+    var kind = kindAndValue.kind
+    var value = kindAndValue.value
+    if (subName != null) {
+      value = OptionElement.create(subName, kind, value)
+      kind = Kind.OPTION
+    }
+    return OptionElement.create(name, kind, value, isParenthesized)
+  }
+
+  /** Reads a value that can be a map, list, string, number, boolean or enum.  */
+  private fun readKindAndValue(): KindAndValue {
+    val peeked = reader.peekChar()
+    when (peeked) {
+      '{' -> return KindAndValue(MAP, readMap('{', '}', ':'))
+      '[' -> return KindAndValue(LIST, readList())
+      '"', '\'' -> return KindAndValue(STRING, reader.readString())
+      else -> {
+        if (Character.isDigit(peeked) || peeked == '-') {
+          return KindAndValue(NUMBER, reader.readWord())
+        }
+        val word = reader.readWord()
+        return when (word) {
+          "true" -> KindAndValue(BOOLEAN, "true")
+          "false" -> KindAndValue(BOOLEAN, "false")
+          else -> KindAndValue(ENUM, word)
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns a map of string keys and values. This is similar to a JSON object, with '{' and '}'
+   * surrounding the map, ':' separating keys from values, and ',' separating entries.
+   */
+  private fun readMap(
+    openBrace: Char,
+    closeBrace: Char,
+    keyValueSeparator: Char
+  ): Map<String, Any> {
+    if (reader.readChar() != openBrace) throw AssertionError()
+    val result = LinkedHashMap<String, Any>()
+    while (true) {
+      if (reader.peekChar(closeBrace)) {
+        // If we see the close brace, finish immediately. This handles {}/[] and ,}/,] cases.
+        return result
+      }
+
+      val option = readOption(keyValueSeparator)
+      val name = option.name
+      val value = option.value
+      if (value is OptionElement) {
+        var nested = result[name] as? MutableMap<String, Any>
+        if (nested == null) {
+          nested = LinkedHashMap()
+          result[name] = nested
+        }
+        nested[value.name] = value.value
+      } else {
+        // Add the value(s) to any previous values with the same key
+        val previous = result[name]
+        when (previous) {
+          null -> result[name] = value
+          is List<*> -> // Add to previous List
+            addToList(previous as MutableList<Any>, value)
+          else -> {
+            val newList = ArrayList<Any>()
+            newList.add(previous)
+            addToList(newList, value)
+            result[name] = newList
+          }
+        }
+      }
+
+      // Discard optional ',' separator.
+      reader.peekChar(',')
+    }
+  }
+
+  /** Adds an object or objects to a List.  */
+  private fun addToList(
+    list: MutableList<Any>,
+    value: Any
+  ) {
+    if (value is List<*>) {
+      list.addAll(value as Collection<Any>)
```

Comment:
Do you still need a cast here? I guess because of the generic...

Reply:
yep